src/TableFunctions/ITableFunctionDataLake.h

```diff
@@ -0,0 +1,93 @@
+#pragma once
+
+#include "config.h"
+
+#if USE_AWS_S3
+
+#    include <TableFunctions/ITableFunction.h>
+#    include <Access/Common/AccessFlags.h>
+#    include <Formats/FormatFactory.h>
+#    include <Interpreters/Context.h>
+#    include <Interpreters/parseColumnsListForTableFunction.h>
+#    include <TableFunctions/TableFunctionS3.h>
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+}
+
+template <typename Name, typename Storage, typename Configuration>
+class ITableFunctionDataLake : public ITableFunction
+{
+public:
+    static constexpr auto name = Name::name;
+    std::string getName() const override
+    {
+        return name;
+    }
+
+protected:
+    StoragePtr
+    executeImpl(const ASTPtr & /*ast_function*/, ContextPtr context, const std::string & table_name, ColumnsDescription /*cached_columns*/)
+        const override
+    {
+        ColumnsDescription columns;
+        if (configuration.structure != "auto")
+            columns = parseColumnsListFromString(configuration.structure, context);
+
+        StoragePtr storage = std::make_shared<Storage>(
+            configuration, StorageID(getDatabaseName(), table_name), columns, ConstraintsDescription{}, String{}, context, std::nullopt);
+
+        storage->startup();
+
+        return storage;
+    }
+
+    const char * getStorageTypeName() const override { return Storage::name; }
+
+    ColumnsDescription getActualTableStructure(ContextPtr context) const override
+    {
+        if (configuration.structure == "auto")
+        {
+            context->checkAccess(getSourceAccessType());
+            return Storage::getTableStructureFromData(configuration, std::nullopt, context);
+        }
+
+        return parseColumnsListFromString(configuration.structure, context);
+    }
+
+    void parseArguments(const ASTPtr & ast_function, ContextPtr context) override
+    {
+        /// Parse args
+        ASTs & args_func = ast_function->children;
+
+        const auto message = fmt::format(
```

Comment:
The message?

Reply:
The `message`?

Reply:
yeah, this will allow to make it not dependant on s3, but also comment from https://github.com/ClickHouse/ClickHouse/pull/45384#discussion_r1108375320 applies here too. It will allow to get rid of multiple templates and will require only one Storage template.